internal/bgp/bgp.go

```diff
@@ -525,8 +527,26 @@ type tcpmd5sig struct {
 // proper TCP MD5 options when the password is not empty. Works by manupulating
 // the low level FD's, skipping the net.Conn API as it has not hooks to set
 // the neccessary sockopts for TCP MD5.
-func dialMD5(ctx context.Context, addr, password string) (net.Conn, error) {
-	laddr, err := net.ResolveTCPAddr("tcp", "[::]:0")
+func dialMD5(ctx context.Context, addr string, srcAddr net.IP, password string) (net.Conn, error) {
+	// If srcAddr exists on any of the local network interfaces, use it as the
+	// source address of the TCP socket. Otherwise, use the IPv6 unspecified
+	// address ("::") to let the kernel figure out the source address.
+	// NOTE: On Linux, "::" also includes "0.0.0.0" (all IPv4 addresses).
+	a := "[::]"
+	if srcAddr != nil {
+		ifs, err := net.Interfaces()
+		if err != nil {
+			return nil, fmt.Errorf("Querying local interfaces: %w", err)
+		}
+
+		if !localAddressExists(ifs, srcAddr) {
```

Comment:
I don't know, but does this check pass with link-local addresses? I guess so, but wanted to double check just in case :)

Reply:
Ha. Indeed we had a bug with IPv6 addresses. Fixed now.